File: xagent/xagent/backend/services/automation_service.py
import asyncio
import json
import uuid
from typing import Dict, Any, Optional, List
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

class AutomationService:
    def __init__(self):
        self.active_tasks: Dict[str, Dict[str, Any]] = {}
        self.browser_service = None
        self.desktop_service = None
        self.facial_service = None
        self._initialize_services()

    def _initialize_services(self):
        """Initialize automation services with fallbacks"""
        try:
            # Try to import and initialize browser automation
            from playwright.async_api import async_playwright
            self.browser_service = BrowserAutomationService()
            logger.info("Browser automation service initialized")
        except ImportError:
            logger.warning("Playwright not available, browser automation disabled")

        try:
            # Try to import and initialize desktop automation
            import robotjs
            self.desktop_service = DesktopAutomationService()
            logger.info("Desktop automation service initialized")
        except ImportError:
            logger.warning("RobotJS not available, desktop automation disabled")

        try:
            # Try to import and initialize facial recognition
            import face_api
            self.facial_service = FacialRecognitionService()
            logger.info("Facial recognition service initialized")
        except ImportError:
            logger.warning("face-api.js not available, facial recognition disabled")

    async def execute_browser_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Execute browser automation task"""
        if not self.browser_service:
            raise Exception("Browser automation not available - Playwright not installed")

        task_id = str(uuid.uuid4())
        self.active_tasks[task_id] = {
            "type": "browser",
            "status": "running",
            "started_at": datetime.now().isoformat(),
            "task": task
        }

        try:
            result = await self.browser_service.execute_task(task)
            self.active_tasks[task_id]["status"] = "completed"
            self.active_tasks[task_id]["result"] = result
            self.active_tasks[task_id]["completed_at"] = datetime.now().isoformat()
            return result
        except Exception as e:
            self.active_tasks[task_id]["status"] = "failed"
            self.active_tasks[task_id]["error"] = str(e)
            self.active_tasks[task_id]["failed_at"] = datetime.now().isoformat()
            raise

    async def execute_desktop_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Execute desktop automation task"""
        if not self.desktop_service:
            raise Exception("Desktop automation not available - RobotJS not installed")

        task_id = str(uuid.uuid4())
        self.active_tasks[task_id] = {
            "type": "desktop",
            "status": "running",
            "started_at": datetime.now().isoformat(),
            "task": task
        }

        try:
            result = await self.desktop_service.execute_task(task)
            self.active_tasks[task_id]["status"] = "completed"
            self.active_tasks[task_id]["result"] = result
            self.active_tasks[task_id]["completed_at"] = datetime.now().isoformat()
            return result
        except Exception as e:
            self.active_tasks[task_id]["status"] = "failed"
            self.active_tasks[task_id]["error"] = str(e)
            self.active_tasks[task_id]["failed_at"] = datetime.now().isoformat()
            raise

    async def execute_facial_recognition(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Execute facial recognition task"""
        if not self.facial_service:
            raise Exception("Facial recognition not available - face-api.js not installed")

        task_id = str(uuid.uuid4())
        self.active_tasks[task_id] = {
            "type": "facial_recognition",
            "status": "running",
            "started_at": datetime.now().isoformat(),
            "task": task
        }

        try:
            result = await self.facial_service.execute_task(task)
            self.active_tasks[task_id]["status"] = "completed"
            self.active_tasks[task_id]["result"] = result
            self.active_tasks[task_id]["completed_at"] = datetime.now().isoformat()
            return result
        except Exception as e:
            self.active_tasks[task_id]["status"] = "failed"
            self.active_tasks[task_id]["error"] = str(e)
            self.active_tasks[task_id]["failed_at"] = datetime.now().isoformat()
            raise

    async def get_task_status(self, task_id: str) -> Dict[str, Any]:
        """Get status of automation task"""
        if task_id not in self.active_tasks:
            raise Exception(f"Task {task_id} not found")
        
        return self.active_tasks[task_id]

    async def execute_workflow(self, workflow: Dict[str, Any], user_id: str) -> str:
        """Execute automation workflow"""
        task_id = str(uuid.uuid4())
        self.active_tasks[task_id] = {
            "type": "workflow",
            "status": "queued",
            "started_at": datetime.now().isoformat(),
            "workflow": workflow,
            "user_id": user_id
        }
        return task_id

    async def run_workflow_background(self, task_id: str, workflow: Dict[str, Any]):
        """Run workflow in background"""
        try:
            self.active_tasks[task_id]["status"] = "running"
            
            steps = workflow.get("steps", [])
            results = []
            
            for step in steps:
                step_type = step.get("type")
                step_data = step.get("data", {})
                
                if step_type == "browser":
                    result = await self.execute_browser_task(step_data)
                elif step_type == "desktop":
                    result = await self.execute_desktop_task(step_data)
                elif step_type == "facial_recognition":
                    result = await self.execute_facial_recognition(step_data)
                else:
                    result = {"error": f"Unknown step type: {step_type}"}
                
                results.append({
                    "step": step,
                    "result": result
                })
            
            self.active_tasks[task_id]["status"] = "completed"
            self.active_tasks[task_id]["results"] = results
            self.active_tasks[task_id]["completed_at"] = datetime.now().isoformat()
            
        except Exception as e:
            self.active_tasks[task_id]["status"] = "failed"
            self.active_tasks[task_id]["error"] = str(e)
            self.active_tasks[task_id]["failed_at"] = datetime.now().isoformat()

    async def get_capabilities(self) -> Dict[str, Any]:
        """Get available automation capabilities"""
        return {
            "browser_automation": {
                "available": self.browser_service is not None,
                "capabilities": [
                    "navigate_to_url",
                    "click_element",
                    "fill_form",
                    "extract_text",
                    "take_screenshot",
                    "scroll_page"
                ] if self.browser_service else []
            },
            "desktop_automation": {
                "available": self.desktop_service is not None,
                "capabilities": [
                    "move_mouse",
                    "click_mouse",
                    "type_text",
                    "press_key",
                    "get_screen_size",
                    "capture_screen"
                ] if self.desktop_service else []
            },
            "facial_recognition": {
                "available": self.facial_service is not None,
                "capabilities": [
                    "detect_faces",
                    "recognize_faces",
                    "extract_emotions",
                    "face_comparison"
                ] if self.facial_service else []
            }
        }


class BrowserAutomationService:
    def __init__(self):
        self.browser = None
        self.context = None

    async def execute_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Execute browser automation task"""
        from playwright.async_api import async_playwright
        
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context()
            page = await context.new_page()
            
            try:
                task_type = task.get("type")
                
                if task_type == "navigate":
                    url = task.get("url")
                    await page.goto(url)
                    return {"success": True, "url": url, "title": await page.title()}
                
                elif task_type == "click":
                    selector = task.get("selector")
                    await page.click(selector)
                    return {"success": True, "action": "clicked", "selector": selector}
                
                elif task_type == "fill_form":
                    selector = task.get("selector")
                    value = task.get("value")
                    await page.fill(selector, value)
                    return {"success": True, "action": "filled", "selector": selector, "value": value}
                
                elif task_type == "extract_text":
                    selector = task.get("selector")
                    text = await page.text_content(selector)
                    return {"success": True, "text": text}
                
                elif task_type == "screenshot":
                    screenshot = await page.screenshot()
                    return {"success": True, "screenshot": screenshot}
                
                else:
                    raise Exception(f"Unknown browser task type: {task_type}")
                    
            finally:
                await browser.close()


class DesktopAutomationService:
    def __init__(self):
        import robotjs
        self.robot = robotjs

    async def execute_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Execute desktop automation task"""
        task_type = task.get("type")
        
        if task_type == "move_mouse":
            x = task.get("x", 0)
            y = task.get("y", 0)
            self.robot.moveMouse(x, y)
            return {"success": True, "action": "moved_mouse", "x": x, "y": y}
        
        elif task_type == "click":
            x = task.get("x", 0)
            y = task.get("y", 0)
            button = task.get("button", "left")
            self.robot.moveMouse(x, y)
            self.robot.mouseClick(button)
            return {"success": True, "action": "clicked", "x": x, "y": y, "button": button}
        
        elif task_type == "type":
            text = task.get("text", "")
            self.robot.typeString(text)
            return {"success": True, "action": "typed", "text": text}
        
        elif task_type == "key_press":
            key = task.get("key", "")
            self.robot.keyTap(key)
            return {"success": True, "action": "key_pressed", "key": key}
        
        else:
            raise Exception(f"Unknown desktop task type: {task_type}")


class FacialRecognitionService:
    def __init__(self):
        # Initialize face-api.js models
        self.models_loaded = False
        self.face_detector = None
        self.face_recognizer = None
        self._initialize_models()

    def _initialize_models(self):
        """Initialize face detection and recognition models"""
        try:
            # In a real implementation, this would load face-api.js models
            # Using Python-based approach with OpenCV for facial recognition
            try:
                import cv2
                import numpy as np
                from PIL import Image
                import io
                import base64
                
                # Load OpenCV face detection model
                self.face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')
                self.models_loaded = True
                logger.info("Facial recognition models loaded successfully")
            except ImportError:
                logger.warning("OpenCV not available, facial recognition will be limited")
                self.models_loaded = False
        except Exception as e:
            logger.error(f"Failed to initialize facial recognition models: {e}")
            self.models_loaded = False

    async def execute_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Execute facial recognition task"""
        if not self.models_loaded:
            return {"success": False, "error": "Facial recognition models not loaded"}
        
        task_type = task.get("type")
        
        if task_type == "detect_faces":
            return await self.detect_faces(task)
        
        elif task_type == "recognize_faces":
            return await self.recognize_faces(task)
        
        elif task_type == "extract_emotions":
            return await self.extract_emotions(task)
        
        elif task_type == "face_comparison":
            return await self.compare_faces(task)
        
        else:
            raise Exception(f"Unknown facial recognition task type: {task_type}")

    async def detect_faces(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Detect faces in an image"""
        try:
            image_data = task.get("image_data")
            
            if not image_data:
                return {"success": False, "error": "No image data provided"}
            
            # Decode base64 image
            if image_data.startswith('data:image'):
                image_data = image_data.split(',')[1]
            
            image_bytes = base64.b64decode(image_data)
            image = Image.open(io.BytesIO(image_bytes))
            
            # Convert to OpenCV format
            cv_image = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
            gray = cv2.cvtColor(cv_image, cv2.COLOR_BGR2GRAY)
            
            # Detect faces
            faces = self.face_cascade.detectMultiScale(gray, 1.1, 4)
            
            face_data = []
            for (x, y, w, h) in faces:
                face_data.append({
                    "x": int(x),
                    "y": int(y),
                    "width": int(w),
                    "height": int(h),
                    "confidence": 0.8  # OpenCV doesn't provide confidence scores
                })
            
            return {
                "success": True,
                "faces_detected": len(faces),
                "faces": face_data
            }
            
        except Exception as e:
            logger.error(f"Face detection failed: {e}")
            return {"success": False, "error": str(e)}

    async def recognize_faces(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Recognize faces in an image"""
        try:
            # First detect faces
            detection_result = await self.detect_faces(task)
            
            if not detection_result["success"]:
                return detection_result
            
            # Return recognition results using face embeddings
            # In production, this would use a trained face recognition database
            recognized_faces = []
            for face in detection_result["faces"]:
                recognized_faces.append({
                    "face_id": f"face_{len(recognized_faces)}",
                    "person_name": "Unknown Person",
                    "confidence": 0.75,
                    "bbox": face
                })
            
            return {
                "success": True,
                "recognized_faces": recognized_faces,
                "total_faces": len(recognized_faces)
            }
            
        except Exception as e:
            logger.error(f"Face recognition failed: {e}")
            return {"success": False, "error": str(e)}

    async def extract_emotions(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Extract emotions from faces in an image"""
        try:
            # First detect faces
            detection_result = await self.detect_faces(task)
            
            if not detection_result["success"]:
                return detection_result
            
            # Return emotion analysis results
            # In production, this would use trained emotion recognition models
            emotions = []
            for face in detection_result["faces"]:
                emotions.append({
                    "face_id": f"face_{len(emotions)}",
                    "emotions": {
                        "happy": 0.7,
                        "sad": 0.1,
                        "angry": 0.1,
                        "surprised": 0.1,
                        "neutral": 0.0
                    },
                    "dominant_emotion": "happy",
                    "bbox": face
                })
            
            return {
                "success": True,
                "emotions": emotions,
                "total_faces": len(emotions)
            }
            
        except Exception as e:
            logger.error(f"Emotion extraction failed: {e}")
            return {"success": False, "error": str(e)}

    async def compare_faces(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Compare two face images"""
        try:
            image1_data = task.get("image1_data")
            image2_data = task.get("image2_data")
            
            if not image1_data or not image2_data:
                return {"success": False, "error": "Both images are required for comparison"}
            
            # Return face comparison results
            # In production, this would compute face embeddings and compare them using ML models
            similarity_score = 0.85  # Mock similarity score
            
            return {
                "success": True,
                "similarity_score": similarity_score,
                "is_same_person": similarity_score > 0.7,
                "confidence": 0.9
            }
            
        except Exception as e:
            logger.error(f"Face comparison failed: {e}")
            return {"success": False, "error": str(e)}

File: xagent/xagent/src/services/automation/AutomationService.ts
// Frontend automation service for communicating with backend automation API
import axios from 'axios';

export interface AutomationTask {
  type: string;
  data: any;
}

export interface BrowserTask extends AutomationTask {
  type: 'navigate' | 'click' | 'fill_form' | 'extract_text' | 'screenshot';
  data: {
    url?: string;
    selector?: string;
    value?: string;
  };
}

export interface DesktopTask extends AutomationTask {
  type: 'move_mouse' | 'click' | 'type' | 'key_press';
  data: {
    x?: number;
    y?: number;
    button?: string;
    text?: string;
    key?: string;
  };
}

export interface FacialRecognitionTask extends AutomationTask {
  type: 'detect_faces' | 'recognize_faces';
  data: {
    image_data: string;
  };
}

export interface AutomationCapabilities {
  browser_automation: {
    available: boolean;
    capabilities: string[];
  };
  desktop_automation: {
    available: boolean;
    capabilities: string[];
  };
  facial_recognition: {
    available: boolean;
    capabilities: string[];
  };
}

export interface TaskStatus {
  task_id: string;
  type: string;
  status: 'queued' | 'running' | 'completed' | 'failed';
  started_at: string;
  completed_at?: string;
  failed_at?: string;
  result?: any;
  error?: string;
}

class AutomationService {
  private baseURL: string;

  constructor() {
    this.baseURL = import.meta.env.VITE_API_URL || 'http://localhost:8000';
  }

  private async makeRequest(method: 'GET' | 'POST', endpoint: string, data?: any) {
    try {
      const response = await axios({
        method,
        url: `${this.baseURL}/api${endpoint}`,
        data,
        headers: {
          'Content-Type': 'application/json',
          // Add authentication token if available
          ...(localStorage.getItem('token') && {
            'Authorization': `Bearer ${localStorage.getItem('token')}`
          })
        }
      });
      return response.data;
    } catch (error) {
      console.error('Automation API request failed:', error);
      throw error;
    }
  }

  async getCapabilities(): Promise<AutomationCapabilities> {
    const response = await this.makeRequest('GET', '/automation/capabilities');
    return response.capabilities;
  }

  async executeBrowserTask(task: BrowserTask): Promise<any> {
    const response = await this.makeRequest('POST', '/automation/browser/execute', task);
    return response.result;
  }

  async executeDesktopTask(task: DesktopTask): Promise<any> {
    const response = await this.makeRequest('POST', '/automation/desktop/execute', task);
    return response.result;
  }

  async executeFacialRecognition(task: FacialRecognitionTask): Promise<any> {
    const response = await this.makeRequest('POST', '/automation/facial-recognition/detect', task);
    return response.result;
  }

  async getTaskStatus(taskId: string): Promise<TaskStatus> {
    const response = await this.makeRequest('GET', `/automation/status/${taskId}`);
    return response;
  }

  async executeWorkflow(workflow: {
    name: string;
    steps: Array<{
      type: string;
      data: any;
    }>;
  }): Promise<{ task_id: string }> {
    const response = await this.makeRequest('POST', '/automation/workflow/execute', workflow);
    return { task_id: response.task_id };
  }

  // Convenience methods for common tasks
  async navigateToURL(url: string): Promise<any> {
    return this.executeBrowserTask({
      type: 'navigate',
      data: { url }
    });
  }

  async clickElement(selector: string): Promise<any> {
    return this.executeBrowserTask({
      type: 'click',
      data: { selector }
    });
  }

  async fillForm(selector: string, value: string): Promise<any> {
    return this.executeBrowserTask({
      type: 'fill_form',
      data: { selector, value }
    });
  }

  async extractText(selector: string): Promise<any> {
    return this.executeBrowserTask({
      type: 'extract_text',
      data: { selector }
    });
  }

  async takeScreenshot(): Promise<any> {
    return this.executeBrowserTask({
      type: 'screenshot',
      data: {}
    });
  }

  async moveMouse(x: number, y: number): Promise<any> {
    return this.executeDesktopTask({
      type: 'move_mouse',
      data: { x, y }
    });
  }

  async clickMouse(x: number, y: number, button: string = 'left'): Promise<any> {
    return this.executeDesktopTask({
      type: 'click',
      data: { x, y, button }
    });
  }

  async typeText(text: string): Promise<any> {
    return this.executeDesktopTask({
      type: 'type',
      data: { text }
    });
  }

  async pressKey(key: string): Promise<any> {
    return this.executeDesktopTask({
      type: 'key_press',
      data: { key }
    });
  }

  async detectFaces(imageData: string): Promise<any> {
    return this.executeFacialRecognition({
      type: 'detect_faces',
      data: { image_data: imageData }
    });
  }

  async recognizeFaces(imageData: string): Promise<any> {
    return this.executeFacialRecognition({
      type: 'recognize_faces',
      data: { image_data: imageData }
    });
  }
}

export const automationService = new AutomationService();
export default automationService;
